### ISSUE TITLE
deprecated Resource subclasses in twisted.web.error, add equivalents in twisted.web.resource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3035]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3035|
|Type|enhancement|
|Created|2008-02-14 19:21:13Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-error-pages-3035|

This circularity is dumb and it breaks tools (eg pydoctor).

Generally, `error` modules are for exceptions, anyway.

<details><summary>Searchable metadata</summary>

```
trac-id__3035 3035
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecate_error_pages_3035 branches/deprecate-error-pages-3035
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1203016873000000 1203016873000000
changetime__1232919622000000 1232919622000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26020]) Branching to 'deprecate-error-pages-3035'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Done, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

  1. A number of tests emit the new warnings, such as:
```
/home/glyph/Projects/Twisted/trunk/twisted/web/_auth/wrapper.py:120: DeprecationWarning: twisted.web.error.ErrorPage is deprecated since Twisted 9.0.  See twisted.web.resource.ErrorPage.
```
  1. The moved resources should get docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Replying to [glyph](#note_4):
>   1. A number of tests emit the new warnings, such as:
> ```
> /home/glyph/Projects/Twisted/trunk/twisted/web/_auth/wrapper.py:120: DeprecationWarning: twisted.web.error.ErrorPage is deprecated since Twisted 9.0.  See twisted.web.resource.ErrorPage.
> ```

Okay, I think I replaced all uses of the deprecated API with the non-deprecated API.  Running the test suite like this:

```
PYTHONPATH= python -W 'error:twisted.web.error.ErrorPage is deprecated since Twisted 9.0.  See twisted.web.resource.ErrorPage.' -W 'error:twisted.web.error.NoResource is deprecated since Twisted 9.0.  See twisted.web.resource.NoResource.' -W 'error:twisted.web.error.ForbiddenResource is deprecated since Twisted 9.0.  See twisted.web.resource.ForbiddenResource.' ~/Projects/Twisted/branches/deprecate-error-pages-3035/bin/trial twisted
```

now produces no errors for me.  I had to change a bunch of code with no unit tests, so I added some.  I did **not** add tests for the Quixote stuff.  Changes in r26031.

>   1. The moved resources should get docstrings.

Fixed in r26029.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

 * Some flakes:
```
twisted/web/test/test_cgi.py:7: 'static' imported but unused
twisted/web/test/test_vhost.py:11: 'Resource' imported but unused
twisted/web/twcgi.py:24: 'html' imported but unused
```
 * The new deprecation warnings are missing tests.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

OK I missed some files :). +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26112]) remove unused imports; replace relative imports with absolute imports; update copyrights; add/reformat module docstrings

refs [#3035](https://github.com/twisted/twisted/issues/3035)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [26114]) work around pydoctor bug; markup cannot span multiple lines :/

refs [#3035](https://github.com/twisted/twisted/issues/3035)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [26115]) Skip 403 test on Windows

refs [#3035](https://github.com/twisted/twisted/issues/3035)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Buildbot revealed some minor issues.  I responded to review feedback and fixed the build issues in the revisions above.  Merging now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [26116]) Merge deprecate-error-pages-3035

Author: exarkun
Reviewer: glyph, therve
Fixes: [#3035](https://github.com/twisted/twisted/issues/3035)

Move the error response resource classes from `twisted.web.error` to
`twisted.web.resource`.  This is the first step in eliminating the
circular dependency between those two modules.  Deprecated versions
of the resources remain in `twisted.web.error` for backwards compatibility
and numerous tests for these classes have been added.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
